query-parser/src/parse.rs

```diff
@@ -70,6 +69,130 @@ macro_rules! matches_plain_symbol {
     }
 }
 
+trait FromValue<T> {
+    fn from_value(edn::Value) -> Option<T>;
+}
+
+impl FromValue<PatternNonValuePlace> for PatternNonValuePlace {
+    fn from_value(v: edn::Value) -> Option<PatternNonValuePlace> {
+        match v {
+            edn::Value::Integer(x) => if x >= 0 {
+                Some(PatternNonValuePlace::Entid(x as u64))
+            } else {
+                None
+            },
+            edn::Value::PlainSymbol(x) => if x.0.as_str() == "_" {
+                Some(PatternNonValuePlace::Placeholder)
+            } else if x.is_var_symbol() {
+                Some(PatternNonValuePlace::Variable(Variable(x)))
+            } else {
+                None
+            },
+            edn::Value::NamespacedKeyword(x) =>
+                Some(PatternNonValuePlace::Ident(x)),
+            _ => None,
+        }
+    }
+}
+
+impl FromValue<PatternValuePlace> for PatternValuePlace {
```

Comment:
This is static and strongly typed: one calls `PatternValuePlace::from_value(v)`, and it can only return an `Option<PatternValuePlace>`. Quite possibly it'll be inlined.